pandas/core/reshape/reshape.py

```diff
@@ -1053,6 +1053,168 @@ def get_empty_frame(data) -> DataFrame:
+        cat_columns = []
+        non_cat_columns = []
+        for col in data.columns:
+            if any(ps in col for ps in prefix_sep):
+                cat_columns.append(col)
+            else:
+                non_cat_columns.append(col)
+        data_to_decode = data[cat_columns]
+        non_cat_data = data[non_cat_columns]
+    elif not is_list_like(columns):
+        raise TypeError("Input must be a list-like for parameter 'columns'")
+    else:
+        data_to_decode = data[columns]
+        non_cat_data = data[[col for col in data.columns if col not in columns]]
+
+    # get separator for each prefix and lists to slice data for each prefix
+    if isinstance(prefix_sep, dict):
+        variables_slice = {prefix: [] for prefix in prefix_sep}
+        for col in data_to_decode.columns:
+            for prefix in prefix_sep:
+                if prefix in col:
+                    variables_slice[prefix].append(col)
+    else:
+        sep_for_prefix = {}
+        variables_slice = {}
+        for col in data_to_decode.columns:
+            ps = [ps for ps in prefix_sep if ps in col][0]
+            prefix = col.split(ps)[0]
+            if prefix not in sep_for_prefix:
+                sep_for_prefix[prefix] = ps
+            if prefix not in variables_slice:
+                variables_slice[prefix] = [col]
+            else:
+                variables_slice[prefix].append(col)
+        prefix_sep = sep_for_prefix
+
+    # validate number of passed arguments
+    def check_len(item, name) -> None:
+        if not len(item) == len(variables_slice):
+            len_msg = (
+                f"Length of '{name}' ({len(item)}) did not match the "
+                "length of the columns being encoded "
+                f"({len(variables_slice)})."
+            )
+            raise ValueError(len_msg)
+
+    # obtain prefix to category mapping
+    variables: dict[str, str]
+    if isinstance(variables, dict):
+        check_len(variables, "variables")
+        variables = variables
+    elif is_list_like(variables):
+        check_len(variables, "variables")
+        variables = dict(zip(variables_slice, variables))
+    elif isinstance(variables, str):
+        variables = dict(
+            zip(
+                variables_slice,
+                (f"{variables}{i}" for i in range(len(variables_slice))),
+            )
+        )
+    else:
+        variables = dict(zip(variables_slice, variables_slice))
+
+    if dropped_first:
+        if isinstance(dropped_first, dict):
+            check_len(dropped_first, "dropped_first")
+        elif is_list_like(dropped_first):
+            check_len(dropped_first, "dropped_first")
+            dropped_first = dict(zip(variables_slice, dropped_first))
+        else:
+            dropped_first = dict(
+                zip(variables_slice, [dropped_first] * len(variables_slice))
+            )
+
+    cat_data = {var: [] for _, var in variables.items()}
+    for index, row in data.iterrows():
```

Comment:
Iterating over rows in Python will be too slow - can you have a look at how the (now closed) PR did it?

Reply:
Removed the row iteration. At the moment this resulted in a problem with NaN values in the output DF which I am currently looking into. (Also realised that I forgot to handle cases of NaN values in the input data which is in the process of being solved). I can mirror the method of the old PR if its method is more efficient (or if it provides an easy solution for the NaN issue).